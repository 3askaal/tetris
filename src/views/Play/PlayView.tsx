import React, { useContext, useEffect } from 'react'
import { Box, Container, Wrapper, Popup, Text, Button } from '3oilerplate'
import ReactGA from 'react-ga4'
import { Controls, Map, Score } from '../../components'
import { GameContext } from '../../context'
import { useKeyboardBindings } from '../../helpers/keyboard'

const PlayView = () => {
  const {
    onStartGame,
    gameOver,
    gamePaused,
    setGamePaused,
  } = useContext(GameContext)

  useKeyboardBindings()

  useEffect(() => {
    ReactGA.send({ hitType: "pageview", page: "/play" });
    onStartGame()
  }, [])

  return (
    <Wrapper s={{ p: ['s', 'm', 'l'] }}>
      <Container s={{ p: 0 }}>
        <Box s={{ display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'center', height: '100%' }}>
          <Box s={{ display: 'flex', alignItems: 'flex-end' }}>
            <Score />
          </Box>
          <Map />
          <Box s={{
            display: 'flex',
            width: '100%',
            alignItems: 'flex-end',
            justifyContent: 'center',
            mt: 's'
          }}>
            <Controls />
          </Box>
        </Box>
      </Container>
      { gameOver && (
        <Popup
          actions={[
            <Button data-testid="restart" onClick={() => onStartGame()}>Restart</Button>
          ]}
        >
          <Text s={{ width: '100%', textAlign: 'center' }}>Game over! Click restart to play again.</Text>
        </Popup>
      ) }
      { gamePaused && (
        <Popup
          actions={[
            <Button data-testid="resume" onClick={() => setGamePaused(false)}>Resume</Button>
          ]}
        >
          <Text s={{ width: '100%', textAlign: 'center' }}>Game paused! Click resume to continue playing.</Text>
        </Popup>
      ) }
    </Wrapper>
  )
}

export default PlayView
