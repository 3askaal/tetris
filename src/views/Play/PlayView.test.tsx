import React from 'react';
import "jest-styled-components";
import { last } from "lodash";
import { render, screen, fireEvent, waitFor } from '../../tests';
import "@testing-library/jest-dom";
import PlayView from './PlayView';
import { act } from 'react-dom/test-utils';

let callback: jest.Mock<any, any>

const getLastContext = () => last(callback.mock.calls)[0]

beforeEach(() => {
  callback = jest.fn();
  render(<PlayView />, { callback })

  act(() => {
    callback.mock.calls[0][0].onStartGame()
  })
})

test('moves right', async () => {
  const shapeX = getLastContext().shape.x

  act(() => {
    getLastContext().moveX('right')
  })

  expect(getLastContext().shape.x).toEqual(shapeX + 1);
});

test('moves left', async () => {
  const shapeX = getLastContext().shape.x

  act(() => {
    getLastContext().moveX('left')
  })

  expect(getLastContext().shape.x).toEqual(shapeX - 1);
});

// test('moves left', async () => {
//   // const moveRightButton = screen.getByTestId('move-left')

//   // expect(getLastContext().shape.y).toEqual(2);
//   expect(getLastContext().shape.x).toEqual(9);

//   act(() => {
//     callback.mock.calls[0][0].moveX('left')
//   })

//   expect(callback.mock.calls[2][0].shape.x).toEqual(9);
// });




// describe('Nav', () => {
//   test('Has items', () => {
//     const { getAllByTestId } = render(
//       <Nav to="/page1" items={navItems}>
//         Page 1
//       </Nav>,
//     )

//     expect(getAllByTestId('nav-sub-item').length).toBe(3)
//   })

//   test('Can be closed', () => {
//     const { queryAllByTestId, getByTestId } = render(
//       <Nav to="/page1" items={navItems}>
//         Page 1
//       </Nav>,
//     )

//     const navOpenTrigger: any = getByTestId('nav-open-trigger')

//     fireEvent.click(navOpenTrigger)
//     expect(queryAllByTestId('nav-sub-item').length).toBe(0)

//     fireEvent.click(navOpenTrigger)
//     expect(queryAllByTestId('nav-sub-item').length).toBe(3)
//   })
// })
