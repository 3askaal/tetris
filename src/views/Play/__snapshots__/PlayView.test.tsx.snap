// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainView runs 1`] = `
.c2.c2.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3.c3.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c8.c8.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 1.25rem;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  max-width: 1200px;
}

.c1.c1.c1 {
  padding: 0;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 > * + * {
  margin-left: 0.75rem;
}

.c5.c5.c5 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  padding: 1.25rem;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0.c0.c0 {
  padding: 0.75rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  width: undefinedpx;
  height: undefinedpx;
  margin: auto;
  background-color: #0d0d0d;
  box-shadow: 0 0 0 2px #161616;
}

.c7 {
  position: absolute;
  top: 0px;
  left: 0px;
  width: 0px;
  height: 0px;
  border: .15rem solid;
  border-width: .175rem;
  border-right-color: #fff7ff;
  border-top-color: #fff7ff;
  background-color: #fcc4d1;
  border-left-color: #c893a0;
  border-bottom-color: #c893a0;
  -webkit-transition: all .025s linear;
  transition: all .025s linear;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 100%;
  font-size: .75rem;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 3rem;
  color: #e5e5e5;
  -webkit-flex-basis: 15%;
  -ms-flex-preferred-size: 15%;
  flex-basis: 15%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: .2rem;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  cursor: pointer;
  background-color: #292d4a;
  border-width: .175rem;
  border-style: solid;
  border-top-color: #191e39;
  border-right-color: #191e39;
  border-left-color: #090f29;
  border-bottom-color: #090f29;
  border-radius: .25rem;
  -webkit-transition: all .25s ease;
  transition: all .25s ease;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  pointer-events: none;
}

.c10 svg {
  stroke-width: 2;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 3rem;
  color: #e5e5e5;
  -webkit-flex-basis: 25%;
  -ms-flex-preferred-size: 25%;
  flex-basis: 25%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: .2rem;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  cursor: pointer;
  background-color: #292d4a;
  border-width: .175rem;
  border-style: solid;
  border-top-color: #191e39;
  border-right-color: #191e39;
  border-left-color: #090f29;
  border-bottom-color: #090f29;
  border-radius: .25rem;
  -webkit-transition: all .25s ease;
  transition: all .25s ease;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  pointer-events: none;
}

.c11 svg {
  stroke-width: 2;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 3rem;
  color: #e5e5e5;
  -webkit-flex-basis: 45%;
  -ms-flex-preferred-size: 45%;
  flex-basis: 45%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: .2rem;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  cursor: pointer;
  background-color: #292d4a;
  border-width: .175rem;
  border-style: solid;
  border-top-color: #191e39;
  border-right-color: #191e39;
  border-left-color: #090f29;
  border-bottom-color: #090f29;
  border-radius: .25rem;
  -webkit-transition: all .25s ease;
  transition: all .25s ease;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  pointer-events: none;
}

.c12 svg {
  stroke-width: 2;
}

.c4 {
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  padding-left: 1.25rem;
  padding-right: 1.25rem;
  border-radius: .25rem;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
  background-color: #161616;
  box-shadow: 0 0 0 2px #1d1d1d;
  font-family: monospace,sans-serif;
  font-size: .75em;
}

@media screen and (min-width:320px) {
  .c0.c0.c0 {
    padding: 0.75rem;
  }
}

@media screen and (min-width:576px) {
  .c0.c0.c0 {
    padding: 1.25rem;
  }
}

@media screen and (min-width:320px) {
  .c9 {
    max-width: 22rem;
  }
}

@media screen and (min-width:576px) {
  .c9 {
    max-width: 22rem;
  }
}

@media screen and (min-width:768px) {
  .c9 {
    max-width: 22rem;
  }
}

@media screen and (min-width:320px) {
  .c10 {
    border-width: .175rem;
  }
}

@media screen and (min-width:576px) {
  .c10 {
    border-width: .125rem;
  }
}

@media screen and (min-width:320px) {
  .c11 {
    border-width: .175rem;
  }
}

@media screen and (min-width:576px) {
  .c11 {
    border-width: .125rem;
  }
}

@media screen and (min-width:320px) {
  .c12 {
    border-width: .175rem;
  }
}

@media screen and (min-width:576px) {
  .c12 {
    border-width: .125rem;
  }
}

<div
  className="c0"
  data-testid="wrapper"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <div
            className="c5"
            size="s"
          >
            <span>
              Level: 
              1
            </span>
            <span>
              Score: 
              0
            </span>
            <span>
              Rows: 
              0
            </span>
          </div>
        </div>
      </div>
      <div
        style={
          Object {
            "display": "flex",
            "height": "80%",
            "width": "100%",
          }
        }
      >
        <div
          className="c6"
          style={
            Object {
              "style": undefined,
            }
          }
        >
          <div
            className=""
            data-testid="shape-active"
            shape={
              Object {
                "active": true,
                "blocks": Array [
                  Object {
                    "x": 0,
                    "y": 0,
                  },
                  Object {
                    "x": 0,
                    "y": 1,
                  },
                  Object {
                    "x": 1,
                    "y": 1,
                  },
                  Object {
                    "x": 1,
                    "y": 2,
                  },
                ],
                "color": "#fcc4d1",
                "height": 3,
                "rotated": 3,
                "width": 2,
                "x": 9,
                "y": 2,
              }
            }
            style={
              Object {
                "height": "0px",
                "left": "0px",
                "position": "absolute",
                "top": "0px",
                "width": "0px",
              }
            }
          >
            <div
              className="c7"
              color="#fcc4d1"
              data-testid="shape-active-block"
            />
            <div
              className="c7"
              color="#fcc4d1"
              data-testid="shape-active-block"
            />
            <div
              className="c7"
              color="#fcc4d1"
              data-testid="shape-active-block"
            />
            <div
              className="c7"
              color="#fcc4d1"
              data-testid="shape-active-block"
            />
          </div>
        </div>
      </div>
      <div
        className="c8"
      >
        <div
          className="c9"
        >
          <button
            className="c10"
            data-testid="move-left"
            onClick={[Function]}
            onDragStart={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            type="MOVE"
          >
            <svg
              fill="none"
              height={14}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={14}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="15 18 9 12 15 6"
              />
            </svg>
          </button>
          <button
            className="c11"
            data-testid="rotate"
            onClick={[Function]}
            onDragStart={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            type="SHIFT"
          >
            <svg
              fill="none"
              height={14}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={14}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="23 4 23 10 17 10"
              />
              <path
                d="M20.49 15a9 9 0 1 1-2.12-9.36L23 10"
              />
            </svg>
            <div
              className=""
            >
              SHIFT
            </div>
          </button>
          <button
            className="c12"
            data-testid="drop"
            onClick={[Function]}
            onDragStart={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            type="SPACE"
          >
            <div
              className=""
            >
              SPACE
            </div>
            <svg
              fill="none"
              height={14}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={14}
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="12"
                x2="12"
                y1="5"
                y2="19"
              />
              <polyline
                points="19 12 12 19 5 12"
              />
            </svg>
          </button>
          <button
            className="c10"
            data-testid="move-right"
            onClick={[Function]}
            onDragStart={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            type="MOVE"
          >
            <svg
              fill="none"
              height={14}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={14}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="9 18 15 12 9 6"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
